RFC: Give AbstractArrays smart and performant indexing behaviors for free

mbauman opened the pull request:
~~This is still a work in progress, but I'd like to get some feedback on the architecture and design here before applying this same sort of scheme to `setindex!`.~~

The basic idea is that the only getindex method defined in base for abstract arrays is `getindex(::AbstractArray, I...)`. And the only methods that an AbstractArray subtype must define are `size` and just one `getindex` method:

``` julia
getindex(::T, ::Int) # if linearindexing(T) == LinearFast()
getindex(::T, ::Int, ::Int, #=...ndims(A) indices...=#) if LinearSlow()
```

Unfortunately, it is currently impossible to express the latter method for arbitrary dimensionalities, but in practice it's not a big issue: most LinearSlow() arrays have a fixed dimension.

This is achieved through dispatch on an internal `_getindex` method, which recomputes the indices such that it can call the canonical `getindex` method that the user must define.  If the user has not defined their canonical method, it will fall back to an error method in `_getindex`.  I use similar scheme for `unsafe_getindex`, with the exception that we can fallback to the safe version if the subtype hasn't defined the canonical unsafe method.  This enables fast vector indexing by checking bounds of the index vectors instead of on each element.  And once `@inbounds` is extensible, AbstractArrays will be able to support it by default.

The difficulty with all this redirection is that an extra function call can wreck indexing performance, and it can be hard to avoid.  ~~I've had particular difficulty getting good performance with `CartesianIndexes`, and I still lag in performance there by 20x for big arrays.  I think call site inline annotations would be a magic bullet, but there may be other tricks we can use, too.  I've not looked into this very carefully yet, though.~~ (Fixed with a more sensible inlining strategy)

TL/DR:

In my cursory performance tests [hacked onto Tim's indexing perf suite from his reshape work](https://gist.github.com/mbauman/39bb490ce2bee52417f1) (more tests are needed), I'm close to matching _or outperforming_ master with `Array` with only these definitions:

``` julia
julia> methods(getindex, (Array, Any...))
3-element Array{Any,1}:
 getindex(A::Array{T,N},i::Int) at array.jl:304
 getindex{T<:Real}(A::Array{T,N},I::Range{T<:Real}) at array.jl:347 # Needed for bootstrap
 getindex(A::AbstractArray{T,N},I...) at abstractarray.jl:492
```

(Of course, in places where we're not quite able to close the gap we can always reinstate the specialized methods. This is just a very useful stress-test of both functionality and performance.)

cc: @timholy

mbauman committed:
Adds methods for getindex and setindex! with Colon

Adds optimized methods for getindex and setindex for Array, BitArray, SubArray with ::Colon.
Adds a simple sparse matrix fallback methods to translate Colons into ranges

Internal method changes:
* Augments Base.index_shape to support ::Colon
* Adds a Base.index_lengths that returns the lengths of each index (without dropping trailing scalars)
* Base.setindex_shape_check now takes the index lengths (instead of the indices themselves)

mbauman commented:
(I was expecting the tests to fail, but I'm amazed at the variety of ways they did so.  At a minimum, this needs both #9607 and #10337).

Jutho commented on base/multidimensional.jl:
```diff
@@ -178,100 +172,164 @@ end  # IteratorsMD
+    # Both index_shape and checksize compute sum(I); manually hoist it out
+    N = sum(I)
+    dest = similar(src, (N,))
+    size(dest) == (N,) || throw(DimensionMismatch())
```
why is this size check necessary? Haven't you just created dest of that size?

mbauman replied:
Not really - `similar` has.  It could give us the wrong answer, and the check here is very cheap compared to the allocations and many assignments (although I may need to hide it in a function; I haven't profiled yet).  In general, my approach has been to not trust the output of `similar`. Master effectively does the same thing for arrays (although it's broken across method boundaries): https://github.com/JuliaLang/julia/blob/master/base/multidimensional.jl#L219

timholy commented:
I'm super-excited about this, and will try to review soon. I just have pretty limited time slots right now.

johnmyleswhite commented:
This would be so awesome if it works out.

timholy commented on base/abstractarray.jl:
```diff
@@ -972,28 +1045,17 @@ function repmat(a::AbstractVector, m::Int)
+sub2ind(dims::(Integer,Integer), i1::Integer) = i1
+sub2ind(dims::(Integer,Integer), i1::Integer, i2::Integer) = i1+dims[1]*(i2-1)
+sub2ind(dims::(Integer,Integer), i1::Integer, i2::Integer, I::Integer...) = i1+dims[1]*(i2-1+sum(I)-length(I))
+sub2ind(dims::(Integer,Integer,Integer...), i1::Integer, i2::Integer, I::Integer...) =
```
Do you need an `@inline` here to ensure good performance for higher dimensions?

mbauman replied:
Yes, for N>5. See https://github.com/JuliaLang/julia/pull/10337#issuecomment-76300541.

timholy commented on base/multidimensional.jl:
```diff
@@ -178,100 +172,164 @@ end  # IteratorsMD
+# Indexing with just one array is always linear in the source, fast or slow
+# TODO: Use magic to speed up repeated ind2sub calls
+@inline function _unsafe_getindex!(::LinearFast, dest::AbstractArray, ::LinearIndexing, src::AbstractArray, I::AbstractArray)
+    for i=1:length(dest)
```
How about a construct like this:
```julia
idest = 0
for isrc in I
    idest += 1
    unsafe_setindex!(dest, getindex(src, isrc), idest)
end
```
That way you don't implicitly assume that `I` has fast linear indexing.

mbauman replied:
Ah, yes, very good idea.

timholy commented on base/multidimensional.jl:
```diff
@@ -178,100 +172,164 @@ end  # IteratorsMD
 
-# Also a safe version of getindex!
-stagedfunction getindex!(dest, src, I::Union(Real,AbstractVector)...)
+checksize(A::AbstractArray, I::AbstractArray) = true # reshaped linear idx
```
How do you prevent this from being called directly when there's only one index?

mbauman replied:
Ah, good catch.  I think this is leftover from an attempted refactor of `checksize` that didn't use stagedfunctions.  It's incorrect in either case, though.

timholy commented on base/operators.jl:
```diff
@@ -248,7 +261,7 @@ function setindex_shape_check(X::AbstractArray, I...)
     while true
         ii = size(X,i)
-        jj = length(I[j])::Int
+        jj = I[j]
```
Since this function used to take the indexes themselves rather than their sizes, I wonder if we should change the signature to `I::Int...`. There might have been some callers outside of base.

mbauman replied:
Yes, that seems sensible.

timholy commented:
This is awesome. In broad terms, I agree this is how indexing should work. You also have a lot of clever insights into how to implement this---it was fun to read your code.

You're basically building this on top of #8227? Perhaps we should just merge that and deal with the consequences later.

> Unfortunately, it is currently impossible to express the latter method for arbitrary dimensionalities, but in practice it's not a big issue: most LinearSlow() arrays have a fixed dimension.

I'm not sure I agree with this. ReshapedArrays.jl, Interpolations.jl, and your own AxisArrays.jl (if you pass in a `LinearSlow()` parent) are good counter-examples. I fear we need to introduce `getindexN`.

Finally, scalar indexing with non-reals is a genuine possibility, see the remarks re `DualNumbers.jl` in #9586.

mbauman commented:
Thanks!

> You're basically building this on top of #8227?

Yes, in spirit.  I just needed some way to express `getindex` without a bounds check; BitArrays and a few other data structures already use this idiom, too.  I really want user-extensible `@inbounds`, too, but after working on this I'm not sure having two independent methods is the best solution.  It adds a lot of indirection here, and it'd be nice if it could work for things like `sub2ind`.

I think your alternative proposal in https://github.com/JuliaLang/julia/issues/7799#issuecomment-50877015 might be more attractive, but that just punts the complexity up the chain to a system I don't know and only a few could implement.  The method table would keep track of the methods compiled with and without bounds checks.  When compiling without bounds checks, the compiler would simply elide everything within `@boundscheck`. I don't think `@inbounds` should propagate, so every performant `getindex` method would be written `getindex(...) = (@boundscheck …; @inbounds …)`.

mbauman commented:
> The method table would keep track of the methods compiled with and without bounds checks.

Hunh, that sounds a lot like multiple dispatch.  `getindex(::Type{BoundsCheckOn}, x, I...)`?  That would actually solve the issues in #8227 with rewriting function names.

simonster commented:
Using multiple dispatch has some elegance, but it has problems of its own.

The first problem is that it makes defining `getindex` uglier for cases where performance doesn't matter: if `getindex(x, I...)` is now `getindex(::Type{BoundsCheckOn}, x, I...)`, then it seems that all existing definitions need to be changed. We could have a deprecation that gets thrown when an old `getindex` method is called, similar to @vtjnash's deprecation for `convert` with pointer types, but arguably users shouldn't need to think about bounds checks unless they need to turn them off for performance purposes.

The second problem is that, at least as far as I can tell, there's still a lowering issue. It's not possible to create an `@inbounds` macro that changes the first argument to `getindex` without modifying the frontend (or reproducing the lowering code in Julia), since `ref` is not lowered to `getindex` in the AST that the macro gets. Using multiple dispatch would solve @JeffBezanson's scoping concerns, but it seems we'd still need to add a new Expr or an argument to `ref`to specify whether bounds checks are on or off.

timholy commented:
Following up from https://github.com/JuliaLang/julia/pull/10507#issuecomment-83846919. @JeffBezanson, your input here would be helpful: how can we restrict dispatch to calling a particular method only for a specific number of varargs? I see a couple of options:
- Introduce `getindexN` as an indexing function that only gets called when `length(indexes) == ndims(A)`
- https://github.com/JuliaLang/julia/pull/10507#issuecomment-84017337
- Introduce `getindex{T,N}(A::AbstractArray{T,N}, I::NTuple{N})` (forces a tuple creation unless inlined)
- A new notation, `getindex{T,N}(A::AbstractArray{T,N}, indexes....N}`.

To me the last seems most attractive, followed by the first or third.

timholy commented:
One more option worth considering: rather than a tuple, use a `CartesianIndex{N}`. The main catch: in order to support things like `InterpolationArray`, the element type should be of any `Number`. This is a little awkward, given that the main mission for `CartesianIndex` is as a glorified counter.

timholy commented:
Barring further suggestions, I'm going to (time permitting) start playing around with adding a new field to a `jl_methlist_t`: an index into the `tvars` field for the parameter specifying the length of the varargs list.
With some parser changes, that should allow us to support the syntax
```julia
getindex{T,N}(A::AbstractArray{T,N}, indexes...N)
```

mbauman commented:
That would be awesome!

mbauman commented:
I've slowly been chipping away at this as I have time.  Setindex is now supported, and I'm getting pretty good performance for almost all my test cases.  SubArrays still need some work, though… their getindex methods should actually be called `unsafe_getindex`, but I don't really want to do that until #8227 is resolved.  I also need to test multidimensional (for > 2 dims) performance more thoroughly.  But it's coming along!

timholy commented:
Yay, that's great! We also probably can't get all the goodness out of this until the tuple work gets completed (consensus discussion on #10691).

tkelman commented:
CI failures were a connectivity issue on travis and #10875 on appveyor, restarted

mbauman commented:
Alright, I've updated again with a few more tweaks, and committed the perf tests I've been using.  The current results are here: https://gist.github.com/mbauman/39bb490ce2bee52417f1

Does anyone have a nice utility to compare the output from the perf tests?

mbauman commented:
Are all six of these CI failures really unrelated?

tkelman commented:
CI is horribly unstable right now, post-tuple change. It's really hard to get a green build in. The 32 bit Travis failures have been seen elsewhere, but the 64 bit ones look like something different. 32 bit appveyor looks new also:
```
exception on 3: ERROR: LoadError: test error in expression: base(3,typemin(UInt64)) == "0"
InexactError()
 in base at intfuncs.jl:283
 in anonymous at test.jl:87
 in do_test at test.jl:47
 in runtests at C:\projects\julia\test\testdefs.jl:77
 in anonymous at task.jl:11
 in anonymous at multi.jl:836
 in run_work_thunk at multi.jl:587
 in anonymous at task.jl:836
while loading numbers.jl, in expression starting on line 275
	From worker 2:	     * core                 in   9.90 seconds
ERROR: LoadError: LoadError: test error in expression: base(3,typemin(UInt64)) == "0"
InexactError()
while loading numbers.jl, in expression starting on line 275
while loading C:\projects\julia\test\runtests.jl, in expression starting on line 3
 
	From worker 3:	     * numbers             Command exited with code 1
```

timholy commented:
@mbauman, I should ask: while I know the full benefit of this PR won't be realized until I finish #10911, it just occurred to me (duh) that perhaps we could merge this anyway: it will help people who write `AbstractArray` types of fixed dimensionality. Is there any conflict you're aware of?

If you think we're good to go, let me know if you want another round of review and I'll see what I can do. (I haven't looked at this in a while, but I also suspect that if you think you're in good shape, then you probably are.)

mbauman commented:
Yes, this is getting into pretty good shape.  The last major issue here is getting `SubArray` to propagate `unsafe_getindex` nicely.  I've been dragging my feet on that one, hoping that #7799 would magically pick up steam and come to a resolution while waiting for #10911.

I'll rebase this (gee, thanks @mbauman for all the merge conflicts from #10884), squash it down, and look into what's needed for `SubArray` to play nicely.

timholy commented:
To clarify, I didn't mean to imply that you needed to get your act together, merely that I didn't want to hold up good work if it's ready to merge. We'll both keep plugging away on our respective tasks.

But exciting to hear it's so close! I agree that #7799 seems likely to be the last barrier to nirvana. (Not what I would have guessed, but it's not an easy problem.)

mbauman commented:
Don't worry, I didn't read that tone into your comment at all.  I'm just as excited to wrap this up myself!

tlycken commented:
I just want to leave a :+1: here for merging this sooner rather than later, even if there are a few other things left to fix before we get all we want out of this; having this in place will make my life a huge lot easier when working on `Interpolations.jl`. As it turns out, I spend more than half my time working on that library just trying to figure out the correct way to avoid ambiguity warnings with `getindex` methods defined in base. If I could just define *one* and be done...

mbauman commented:
Alright, updated with SubArray indexing refactored to play nicely with the fallbacks.  Indexing SubArrays now checks bounds, and indexes into the parent with `unsafe_getindex`.  `JULIA_TESTFULL` passes on my machine.

I think this is ready for a review.  I'd particularly appreciate performance tests on Array, SubArray (cc: @timholy), and BitArray (cc: @carlobaldassi) to ensure there aren't major regressions.  Using the attached [perf suite (on just those types)](https://gist.github.com/mbauman/d422fe3b3f35b7193748), I'm improving the performance of Array across the board (!!).

SubArray is more mixed, but note that `for I in subArr` iteration is now checking bounds (since since it's using next).  Generally within 10%.

BitArray is also more mixed.  Generally within 20%.

Any tips on how I can improve things here would be greatly appreciated.

tkelman commented:
> AbstractArrays that do not extend `similar` now return an `Array` by default

What are we doing for this now?

mbauman commented:
> What are we doing for this now (prior to this PR)?

Throwing a no method error. Since indexing leans heavily on `similar` this seemed sensible.

IainNZ commented:
@mbauman this looks really neat. Is this expected to be a breaking change in any way?

tkelman commented:
> Throwing a no method error.

Thanks, good to know. I feel like this makes `similar` a pretty vital part of the informal interface of AbstractArray re: https://github.com/JuliaLang/julia/issues/10064 so should be especially emphasized in the documentation.

mbauman commented:
> Is this expected to be a breaking change in any way?

Nope! It should just add features where there were previously no method errors.

timholy commented:
I didn't look over the perf numbers in great detail (we really need a way of exporting the perf output format to a graphical representation...), but I'm _thrilled_ with a 10% speed decrease on SubArrays in exchange for bounds checking, especially since it sounds like I can turn it off.

Going to start reading through the code (top to bottom) now.

timholy commented on NEWS.md:
```diff
@@ -171,12 +171,23 @@ Library improvements
+      for scalar indices to support indexing; all other indexing behaviors
+      (including logical idexing, ranges of indices, vectors, colons, etc.) are
+      implemented in default fallbacks. Similarly, they only need to implement
+      scalar `setindex!` to support all forms of indexed assingment ([#10525]).
```
Making this true in practice for general types is presumably conditional on merging #10911, but gosh darnit I am _going_ to figure out the remaining challenges for the NTuple/Vararg unification. Status: now building julia only crashes after inference gets turned on, rather than in the 2nd file, which to me seems like progress :smile:.

mbauman replied:
I think I worded this in a way that it's still technically true.  You just need to implement *all* combinations of scalar indexing (linear and otherwise) until #10911.  But, yes, I'm looking forward to it!

timholy commented on base/abstractarray.jl:
```diff
@@ -116,41 +116,53 @@ linearindexing{T<:Range}(::Type{T}) = LinearFast()
+_checkbounds{T<:Real}(sz::Int, r::Range{T}) = (@_inline_meta; _checkbounds(sz, to_index(r)))
+_checkbounds(sz::Int, ::Colon) = true
+function _checkbounds{T <: Real}(sz::Int, I::AbstractArray{T})
+    @_inline_meta
```
Do we really want to force inlining on this?

mbauman replied:
Yes, that is a very good question, and it'd probably be good to review all the little performance tweaks I've been making to make sure they're necessary.  This one, however, makes a huge difference.  It's on the order of 10x more costly to call this function than it is to inline it.  These bounds checks are effectively "inlined" into the current n-ary `arrayref` intrinsics, so I needed to do all sorts of these inlining tricks to match it.  These sorts of optimizations are *so* finicky and tricky (and time consuming!) to test.

I've also played with changing the branching behavior (e.g., using `&` instead of `&&`, etc) and can get really nice-looking (and short!) LLVM code that ends up performing worse for some strange cache/pipeline/who-knows-what reason.

timholy replied:
I really like how you wrote the function---presumably it's SIMD-able this way.

And, thanks for letting me know about the importance of inlining. I assumed anytime you cranked up a for-loop of possibly large size (but, possibly not), it wouldn't be worth inlining. Certainly I wouldn't have guessed this. Thanks for testing!

timholy commented on base/abstractarray.jl:
```diff
@@ -116,41 +116,53 @@ linearindexing{T<:Range}(::Type{T}) = LinearFast()
-function checkbounds(A::AbstractMatrix, I::Union(Real,Colon,AbstractArray), J::Union(Real,Colon,AbstractArray))
-    checkbounds(size(A,1), I)
-    checkbounds(size(A,2), J)
+checkbounds(A::AbstractArray, I::AbstractArray{Bool}) = size(A) == size(I) || _boundserror(A, I)
```
Just noting the inconsistency here (which you inherited) with the method above for `setindex!`.

mbauman replied:
Ah, thanks for the reminder.  I think I'll go back and re-arrange this and deprecate that.

timholy commented on base/abstractarray.jl:
I suspect we want to get rid of the typing on `I` and the calls to `to_index`. In my view, for an `AbstractArray` with LinearSlow, this policy should be handled by the array type itself. Interpolations.jl will thank you for it :smile:.

timholy commented on base/abstractarray.jl:
This call to `to_index` is fine, however.

timholy commented on base/abstractarray.jl:
Would it be worth putting these into a `for (funcname, exprs) in ((:_getindex, :_unsafe_getindex), ...) @eval begin ...` block?

timholy commented on base/array.jl:
```diff
@@ -279,112 +279,62 @@ done(a::Array,i) = (i > length(a))
 
 ## Indexing: getindex ##
 
-getindex(a::Array) = arrayref(a,1)
```
Gotta leave now, but I'll get back to this hopefully later today.

mbauman replied:
(This is so great, thank you!)

timholy commented on base/array.jl:
```diff
@@ -279,112 +279,62 @@ done(a::Array,i) = (i > length(a))
+#         unsafe_copy!(X, 1, A, 1, lI)
+#     end
+#     return X
+# end
```
I assume you still want these commented out, but a reminder about your TODO

timholy commented on base/multidimensional.jl:
```diff
@@ -222,117 +163,249 @@ end  # IteratorsMD
+        # This is specifically *not* inlined.
+        @nexprs $N d->(I_d = to_index(I[d]))
+        dest = similar(A, @ncall $N index_shape A I)
+        @ncall $N checksize dest I
```
What's this `checksize` for?

mbauman replied:
It's to ensure that the array returned from `similar` is the size you asked for.  While we're putting a lot of faith into user types and definitions with all these `unsafe_` methods, this check is very cheap compared to the non-scalar indexing and allocation.  It's actually something we were already doing in some of the Array code, and I thought it was prudent to carry forward.

timholy replied:
Good point, I didn't think about what would happen if similar is borked for that particular array type. If the test is cheap, definitely worth having.

timholy commented on base/multidimensional.jl:
```diff
@@ -222,117 +163,249 @@ end  # IteratorsMD
+            unsafe_setindex!(dest, src[s], d)
+        end
     end
+    dest
```
:+1:

timholy commented on base/multidimensional.jl:
```diff
@@ -222,117 +163,249 @@ end  # IteratorsMD
+# and ensure the value to set is either an AbstractArray or a Repeated scalar
+# before redispatching to the _unsafe_batchsetindex!
+_iterable(v::AbstractArray) = v
+_iterable(v) = repeated(v)
```
Nice

mbauman commented on base/abstractarray.jl:
I actually had that at one point, but decided it'd be better to start verbosely and then condense later if it ends up making sense.  I'll play with it.

timholy commented on base/multidimensional.jl:
```diff
@@ -222,117 +163,249 @@ end  # IteratorsMD
+# 1-d logical indexing: override the above to avoid calling find (in to_index)
+function _unsafe_setindex!(::LinearIndexing, A::AbstractArray, x, I::AbstractArray{Bool})
+    X = _iterable(x)
+    setindex_shape_check(X, index_lengths(A, I)[1])
```
I guess this is `unsafe` only in its accesses of `A`?

timholy commented on base/range.jl:
```diff
@@ -342,12 +342,12 @@ done(r::UnitRange, i) = i==oftype(i,r.stop)+1
-    convert(T, first(r) + (i-1)*step(r))
+    unsafe_getindex(r, i)
 end
+unsafe_getindex{T}(v::Range{T}, i::Integer) = convert(T, first(v) + (i-1)*step(v))
```
I've been looking forward to being able to skip bounds checks on Ranges for a long time. Yay!

timholy commented on base/sparse/sparsematrix.jl:
```diff
@@ -1236,6 +1234,12 @@ end
 getindex{T<:Integer}(A::SparseMatrixCSC, I::AbstractVector{T}, j::Integer) = getindex(A,I,[j])
 getindex{T<:Integer}(A::SparseMatrixCSC, i::Integer, J::AbstractVector{T}) = getindex(A,[i],J)
 
+# Colon translation (this could be done more efficiently)
```
The real problem is anyone _using_ these methods. But yes, someone (not you) should eventually build faster implementations.

timholy commented on base/subarray2.jl:
```diff
@@ -1,107 +1,47 @@
+    if ni == 1 && length(IV.parameters) == LD  # linear indexing
+        meta = Expr(:meta, :inline)
+        if iscontiguous(V)
+            return :($meta; V.parent[V.first_index + I[1] - 1])
```
Just to check: now we only get here if `I` has length 1, right?

mbauman replied:
Missed this comment: yes, this is largely the same as the existing code, just not indented within an `@eval` block anymore.

timholy commented:
Wow. :100:!

timholy commented:
This also scores big-time on an important @JeffBezanson metric: if you discount all those lines of perf tests you're adding, it's a net reduction of about 130 lines of code.

mbauman commented:
Ah, of course, that's where those lines came from.  I had been keeping a close eye on that metric and forgot about the perf tests.  (Although this fails on a few other Jeff-Tests, leaning heavily on `@generated` and `@inline`).

Jutho commented on base/abstractarray.jl:
Would the `_boundserror` call also be useful here, or the GC frame not matter here? I don't know anything about this, so just guessing...

timholy commented on base/multidimensional.jl:
```diff
@@ -222,117 +163,249 @@ end  # IteratorsMD
-        @nexprs $N d->(size(A, d) == length(I[d]) || throw(DimensionMismatch("index $d has length $(length(I[d])), but size(A, $d) = $(size(A,d))")))
-        nothing
+        # This is specifically *not* inlined.
+        @nexprs $N d->(I_d = to_index(I[d]))
```
Forgot to note that we should get rid of this `to_index`, too.

mbauman commented on base/abstractarray.jl:
I definitely agree, but I want to hear more (from you or @tlycken) about how Interpolations would like to use Dual or other non-Real indices. I think these definitions are correct, but that we'll want to change the nonscalar `Union(Real, AbstractVector, Colon)` to be more accepting. 

Does it make sense to use linear indexing with non-real numbers? They would still need to implement arithmetic and div/rem for ind2sub and sub2ind, so perhaps these definitions could be widened to `Number`, but I hesitate to go further. 

For nonscalar indexing, right now all indices must support Base.index_shape, index_lengths, and unsafe_getindex. The tricky one here is indexing - we could have the shape and length default to 1. We could also make to_index default to being a no-op for unknown types.

mbauman commented on base/abstractarray.jl:
Great catch, I forgot about these when I was testing my changes to `checkbounds`! I'll have to test it.

carlobaldassi commented on base/multidimensional.jl:
```diff
@@ -619,91 +648,40 @@ getindex{T<:Real}(B::BitArray, I0::UnitRange{T}) = getindex(B, to_index(I0))
+# in the general multidimensional non-scalar case, can we do about 10% better
+# in most cases by manually hoisting the bitarray chunks access out of the loop
+# (This should really be handled by the compiler or with an immutable BitArray)
+@generated function _unsafe_getindex!(X::BitArray, ::LinearFast, B::BitArray, I::Union(Int,AbstractVector{Int},Colon)...)
```
This is not going to be invoked on generic `AbstractVector` right? This was previously handled in the following (now deleted) method which called `to_index`.
Unless I'm wrong, we would need to add `to_index` calls here when reading `I`, but I don't know about the effect on performance.

mbauman replied:
This fits into the general schema I've set up for all `AbstractArrays`, which jumps through the following functions:

* `getindex(::AbstractArray, ::Any...)` - simply call `linearindexing` and redispatch to
* `_getindex(::LinearIndexing, ::AbstractArray, ::Union(Real, AbstractVector, Colon))` - `checkbounds` and call:
* `_unsafe_getindex(::LinearIndexing, ::AbstractArray, ::Union(Real, AbstractVector, Colon))` - create output array, double-check its size, and call `to_index`, and then dispatch to
* `_unsafe_getindex!` (this function).

So, yes, this could be called with an `AbstractVector`, but it's already passed through `to_index` and its element type is restricted to be `Int`.

carlobaldassi replied:
Ah indeed, sorry for the noise.

mbauman replied:
Not noise at all.  It's good to have more folks who know this code than just me and Tim.  I should add more comments like that into the code itself.

carlobaldassi commented on base/multidimensional.jl:
```diff
@@ -619,91 +648,40 @@ getindex{T<:Real}(B::BitArray, I0::UnitRange{T}) = getindex(B, to_index(I0))
-        X = BitArray(index_shape($(Isplat...)))
-        Xc = X.chunks
-
+        $(Expr(:meta, :inline))
```
Does inlinig here make a noticeable difference?

mbauman replied:
Yes. I didn't keep notes on how much, but I documented this change in my commit message https://github.com/JuliaLang/julia/commit/2b47b996ba7ce3e004f187162a934679fc292256.

carlobaldassi commented:
Wow! This is amazing.
I still need to test performance, but I just went through the code and I only had two fairly minor comments.

timholy commented on base/abstractarray.jl:
Any time one's using `LinearSlow` and the number of indexes is fewer than the number of dimensions, I think it makes sense to call `to_index` on the final index before calling `ind2sub`. But when the number of indexes matches or exceeds the dimensionality, I think one should leave `to_index` out of it.

Yes, we could broaden that to `Union(Number, AbstractVector, Colon)`. Will that/can that satisfy the constraints you mention in your last paragraph?

mbauman commented:
I've been trying to think about how to best support arbitrary types as indices, and what the minimal set of methods an index must support.  As things stand in this PR, in order for an index to be used as part of a fancy non-scalar indexing expression (ignoring the current signatures), they must implement:

* `to_index`.  This really feels like a crutch to me.  There are really only two worthwhile methods in the context of non-scalar indexing: It converts `Range{T!<:Int}` to an `Int` range (tiny performance gain, but O(1) instead of O(N)),  and it calls `find` on `AbstractArray{Bool}` (change in semantics at a performance *loss*).
* `index_shape` and `index_lengths`. These are easy and can simply be defined to be 1 for all non-Array types.
* `unsafe_getindex`. We're already defining `getindex` on `Number`, so that's an easy "broadening" that we can do in the signatures, but I think ideally these methods would simply expand arrays and pass everything else through untouched.

For scalar indexing, it's much more straight-forward.  The goal here is simply to call `ind2sub`/`sub2ind` and expand/shrink to the supported dimensionality.  `to_index` here makes sense - we can only work with Ints.  Anything else the user wants to support they need to implement themselves.

mbauman commented:
It'd be nice to be able to dispatch on `I::Union(Any, Union(Colon, AbstractArray))...` where there must be at least one argument from the second union.  Otherwise broadening these signatures (even just to `Number`) will likely lead to infinite dispatch loops.

timholy commented:
Presumably calling `find` is mixed in terms of its performance implications. If you are indexing an `m`-by-`n` array with two logicals of length `m` and `n`, respectively, then presumably it's sometimes a big win to use `find`. (It adds `O(m)+O(n)` time but might potentially be saving you `O(m*n)`.) That said, far more common is to index with a boolean array that has the same length as the original array, in which case I agree that calling `find` introduces all sorts of problems (two passes through an array rather than one, forcing you to use linear indexing when the type may not support it well, etc).

But in all cases, I'm fine with the idea of putting the burden on the caller (converting your ranges to `Int` and/or calling `find`) if s/he wants the ultimate performance. Better for the indexing operations to "just do what you told them to": if you said iterate over a `Range{Int8}`, then gosh darnit I'll do that.

So basically I think we can expunge `to_index` for everything except that scalar case you mentioned above.

timholy commented:
> It'd be nice to be able to dispatch on I::Union(Any, Union(Colon, AbstractArray))... where there must be at least one argument from the second union. 

Great minds think alike: #10440. (Rather, whenever you work in an area long enough, you end up wishing for the same things.)

mbauman commented:
We could work around the dispatch loop since we're already using staged functions — we can just check for arrays or colons and throw an error otherwise.  The only questions that really remain are:

* How to handle bit/bool arrays without `to_index`.
* How to avoid calling getindex on non-array indices in a sensible way.

If we can solve those two, then the definitions in multidimensional could be for `I::Any...`.

carlobaldassi commented:
I started doing some benchmarking on BitArrays stuff. I started from multidimensional logical indexing (3 dimensions). I put the gist (code + results) [here](https://gist.github.com/carlobaldassi/682827226945a3ff5005).

Most tests are roughly the same speed, a couple (`getindex`, indexing an `Array` with an `Array{Bool}` or `BitVector`) have improved (1.2x~1.5x), and some have regressed quite lot (2x), basically all the `setindex!` ones which index an `Array` with an `Array{Bool}` (but not a `BitArray`).

This is with the latest commits on this branch and master.

mbauman commented:
Awesome, thank you.  I'll try to see what I can do for the regressions.

carlobaldassi commented:
> have improved (1.2x~1.5x)

well, I'm not sure how, but that improvement just disappeared in my following tests. In the meanwhile I made sure that the dimensionality is irrelevant, 1d through 4d I get the same results for those tests.

mbauman commented:
Well, as long as they're still on-par with master, I'm very happy.

I think the performance regression was because I was iterating over the `I` array (which checks bounds within `next`) instead of using `eachindex`/`unsafe_getindex`.  I think we can bring those back up to speed without much trouble.

mbauman commented:
I pushed a change, but it didn't seem to have much of an effect.  I'll have to keep digging.

carlobaldassi commented:
Yes those are on par with master. I tried your latest commit, didn't help with the regressions.

I've been testing indexing with `Vector{Int}`. See the [updated gits](https://gist.github.com/carlobaldassi/682827226945a3ff5005).

I found another smaller regression (between 1.14x and 1.3x) in `getindex` indexing a `Vector{T}` with a `Vector{Int}`. The regression is smaller for `T==Float64`, and larger `T==Bool`. You were likely aware of this anyway. The regression disappears if I call `shuffle!` on the indexing vector, so it's something about cache use I suppose?

On the other hand, I also saw some quite massive improvements (even 33x)!

mbauman commented:
Aha! It's the `setindex_shape_check` - that computes `sum(I)` and takes just as long as the indexing itself (hence the 2x regression).  Base gets around this by checking *after* doing the assignments.  I suppose I can do that, too, but I don't like it as much (it mutates before erroring).

carlobaldassi commented:
Ah I see, this also explains why it's faster when indexing with `BitArrays`.

carlobaldassi commented:
So the same logic should probably be applied [here](https://github.com/JuliaLang/julia/blob/b47d3d079d3cc0fabe93ca4de1a940611ec44145/base/bitarray.jl#L463) and [here](https://github.com/JuliaLang/julia/blob/b47d3d079d3cc0fabe93ca4de1a940611ec44145/base/bitarray.jl#L496).

mbauman commented:
BitArrays are already performing this optimization (both in Base, and I carried it over here, too).  I just didn't do it in the general case.

carlobaldassi commented:
I'm not sure I follow. What I meant is that the code I linked above (unchanged in this branch) implements `A[I]=X` with `I::AbstractVector{Bool}` and checks that `X` is long enough while looping over it, and that it was entirely consumed at the end. So it should be changed to check beforehand instead. No?

mbauman commented:
Ah, yes, that's the other solution.  I was just changing the general case to match master.

mbauman commented:
Do you have strong feelings here, @timholy?  Should we check bounds before mutating (like most other setindex! methods) or is the 2x performance gain here worth it?  C.f. https://github.com/JuliaLang/julia/commit/7b60d5013a81962cc9753f75cbc6859d9ea79360

timholy commented:
I don't have strong feelings: on one hand the 2x performance boost is nice, but certainly the typical pattern in master (e.g., `copy!`) is to check before mutating. I probably decided to go with the performance, reasoning that you'll get an error and so have to rerun your computations anyway.

mbauman commented:
Let's match the algorithm on master for now.  I *really* don't want any performance regressions from this branch, intentional or not. We can make that change separately, but given that nobody has complained in the past year it may be just fine.

StefanKarpinski commented:
@mbauman, your follow-through on this is really impressive.

mbauman commented:
Alright, well, I don't see an immediate solution for supporting `Number` or other such non-real indices in a simple way.  I had forgotten about a very important method that all indices must support for non-scalar indexing: `checkbounds`.  We can't just ignore this one, since we then go on to call unsafe methods with those indices.  If we refactored `checkbounds` so there's a nicer interface to override, would there be a sensible definition for, e.g., Dual Numbers to implement?

timholy commented:
Ooh, arg. I didn't think of that. Rats.

I _guess_ we could call `real` (`DualNumbers` are literally a real value plus an infinitesimal epsilon, so `real` makes sense). But if supporting `Number` is troublesome, perhaps it's best to insist that this only be done via scalar indexing, in which case the user's method `getindex{T,N}(A::MyArray{T,N}, indexes::Vararg{Number,N})` will handle it anyway. (In other words, maybe I sent you on a wild goose chase. Sorry!)

mbauman commented:
I think it's good to consider what the possibilities are here to make this as useful as possible.  We can always try to widen these definitions later, too.

This is looking pretty good.  I think the final holdup here is #10337 and the performance of indexing with the "wrong" dimensionality (e.g., 2+ dims for LinearFast).

mbauman committed:
AbstractArray: fancy getindex and setindex! fallbacks

This implements "fancy" indexing behaviors for all AbstractArray subtypes.
* Converts CartesianIndices to their full dimensionality, regardless of the types of any other indices
* For linear fast arrays, computes the linear index from multidimensional scalars
* For linear slow arrays, computes the full dimensionality indexing subscript from a linear index
* Implements non-scalar indexing:
    * Logical indexing
    * Vectors or ranges of indices
    * Indexing with one multidimensional array of indices

The non-scalar indexing relies heavily on `similar`, so a default implementation for AbstractArrays was created to return an Array.

Additionally, this expands the usage of `Base.unsafe_getindex` for indexing without bounds checks.

mbauman committed:
Adapt AbstractArray subtypes for indexing fallbacks

This removes redundant behaviors from existing AbstractArray subtypes.

mbauman commented:
Alright!  I squashed and rebased (a few times last night) and apparently I've broken GitHub's commit display here in the PR - they're not showing in the correct order anymore.  The [branch view](https://github.com/JuliaLang/julia/compare/mb/abstractsmarts) matches what's in Git.

Only two issues are left:

1. This branch seems to be particularly prone to triggering an EOF error while running the tests - see the current failure on Travis.  This does already happen on master (e.g., https://travis-ci.org/JuliaLang/julia/jobs/61086284 and https://travis-ci.org/JuliaLang/julia/jobs/60371436), but it doesn't happen nearly as often as I've seen it occur here.

2. I've added explicit linear and cartesian performance tests to test the non-optimal codepaths.  Everything looks dandy except one case: `eachindex(A)` is 4x slower than `1:length(A)` for a custom LinearFast array type.  In looking at the `code_typed` output, the two are identical except that `eachindex` has an additional and superfluous `call((top(getfield))(Base,:LinearFast))::Base.LinearFast` in the loop (which should be a noop).  This adds an extra branch for an `jl_undefined_var_error` in the LLVM IR.  It only seems to occur in types that are defined outside of base.  This isn't a deal breaker, but it'd be nice to resolve.

timholy commented:
Is the `EOFError` always in the `parallel` tests? And, can you trigger it locally? Since it's likely due to an independent problem, it could be useful to merge this to make it easier to track down :wink:.

mbauman commented:
Nope, it's happened in a variety of tests.  It does seem to always occur towards the end, though — examples, cmdlineargs, parallel, etc.  I've not seen it locally, but I've not run the full test suite that many times.

mbauman commented:
With regards to point 2, the extra overhead goes away if I `import Base: LinearFast` instead of using the fully-qualified `Base.LinearFast()` in the [linearindexing definition](https://github.com/JuliaLang/julia/pull/10525/files#diff-43993f85f1b636f373d938d73f8f682fR146).  That's a decent (albeit surprising) workaround.

mbauman committed:
Perf tests for indexing

mbauman committed:
News for #10525

mbauman commented:
I rebased over some of the recent GC root fixes, and for the first time that I've noticed this brach got 4/4 :white_check_mark: on Travis!  I'll push again when there's less pressure on the CI backlog to make sure this wasn't just a fluke, but I'm hopeful!

timholy commented:
That's great. It's quite likely that #11190 makes a difference here.

tkelman commented:
BoundsErrors on Win32 unfortunately? https://ci.appveyor.com/project/StefanKarpinski/julia/build/1.0.4547/job/50t4x9w7bqqoq4lv

tkelman commented:
I can reproduce that locally, this must be a codegen bug or something, since printing `rem(x,b)` right before this line https://github.com/JuliaLang/julia/blob/00e9babb726c64507689843cc9454e19182f2d1e/base/intfuncs.jl#L285 makes it go away. I'll also give LLVM 3.6 a try.

tkelman commented:
It looks like this is fixed on LLVM 3.6.0, but we've got some other bugs still open there. Any ideas for workarounds, LLVM experts? cc @vtjnash @Keno 

`base(3, typemin(UInt64), 1, false)` is failing on win32 with LLVM 3.3 on this branch, here's the lowered, llvm, and native code: https://gist.github.com/17f8c8b2a3212942d43d

tkelman commented:
I feel like this issue is on the critical path towards 0.4.0, and this odd, seemingly unrelated codegen bug appears to be the only thing holding it up. Looks like this needs a rebase again, worth trying to see if the codegen bug maybe goes away?

mbauman commented:
Done.  Thanks so much for checking out the x86 Windows failure, Tony.  I hope there's a simple workaround to be found.

tkelman commented:
It looks like appveyor may be broken temporarily by 460300645623d0fbafd4c4f26eded8626ae69019, so that'll need to be reverted or done differently

mbauman commented:
I tried omitting it but AppVeyor didn't take kindly to that, either.  I'll just wait for a fix and push again later.

tkelman commented:
Sadly the codegen problem doesn't look like it's going to go away so easily. I tried a few things using divrem, have yet to find a non-printing workaround to avoid the boundserror. In need of some brainstorming here.

mbauman commented:
It doesn't look like we got rid of that EOFError, either.  This is still hitting it more frequently than master is.  I'm at a loss in debugging either of these issues.

tkelman commented:
Worth trying to rebase this?

johnmyleswhite commented:
Can I request that we add more documentation for this if it's going to be merged in time for 0.4? I tried to use this for NullableArrays and found it more difficult than I expected to determine which minimal methods need to be implemented to get something working.

IainNZ commented:
I've been thinking of an "interfaces" page in the manual that documents our informal interfaces. E.g. that you need start done next for iteterable, isless (?) for sorting, and then ... for indexing.

StefanKarpinski commented:
+1 to both more docs and an interfaces chapter. I do think that we can merge the code before adding the docs as long as we actually add the docs too.

mbauman commented:
Definitely. It's great to have others kicking the wheels here - would you be willing to post your attempt so far?  I'd be willing to help you get it together.

In the meantime, here's a quick memory-dump of what an AbstractArray needs to fully support indexing with this PR (without testing).  I think all you should need is: `size`, `getindex(X::MyArray, I::Int...)` and `setindex!(X::MyArray, I::Int...)`.  `similar` now defaults to giving you back an `Array`, change that only if you need something different.  If your type allows for fast linear indexing, you should opt into `Base.linearindexing{T<:MyArray}(::Type{T}) = LinearFast()` (you should explicitly `import Base: LinearFast` instead of defining it with the module qualifier).

tkelman commented:
Note that the codegen problem on win32 (https://github.com/JuliaLang/julia/pull/10525#issuecomment-100420099) is likely to still be present, unless we happen to get lucky and it goes away after a rebase. That still may need to be debugged before even the code here can be merged.

mbauman commented:
@tkelman I have a Win64 machine here I can try to get up and running.  Is there a list of what all is needed to ensure a 32 bit build?

tkelman commented:
@mbauman follow the cygwin instructions at https://github.com/JuliaLang/julia/blob/master/README.windows.md#cygwin-to-mingw-cross-compiling - set `XC_HOST = i686-w64-mingw32` for a 32 bit build

mbauman commented:
Thanks!

But… has it magically been resolved??  The numbers test passed and now we're hitting a failure in sparse on all platforms (looks like I goofed in my merge resolution).

mbauman commented:
… and SourceForge is down (but instead of acting like it's down it returning a bunch of 302 redirects to a 200 OK page that requires javascript to function). Lovely.

The win32 build squeaked through just in time and managed to pass without any codegen issues once again.

tkelman commented:
Looks like we might actually be okay now. Sourceforge does suck as usual, someone should try to convince PCRE's author to move to github. Had to fiddle with AppVeyor a bit now that we have nightlies with PCRE2 built in them, but assuming the currently queued build goes through, is this good to merge then? I guess we'll have to keep our fingers crossed that the codegen bug doesn't reappear later.

timholy commented:
(Where's the :looking_oh_so_hopeful: emoji??)

mbauman commented:
!!! Everything passed!  I'll clean up the history, and then this will be good to merge after a second round of CI.

tkelman commented:
And the requested doc additions, either now or as a TODO-soon

johnmyleswhite commented:
@mbauman: Thanks for the guidelines. Where I got stuck last time I played with this still confuses me.

I think my confusion lies in some assumptions made by `similar` and its relationship to the implicit `eltype` of an `AbstractArray`. The following example should show where I got derailed:

```
immutable NullableArray{T, N} <: AbstractArray{Nullable{T}, N}
    values::Array{T, N}
    isnull::Array{Bool, N}
end

function NullableArray(T, dims::Dims)
    NullableArray(
        Array(T, dims),
        fill(false, dims)
    )
end

Base.size(X::NullableArray) = size(X.values)

function Base.getindex{T, N}(X::NullableArray{T, N}, I::Int...)
    if X.isnull[I...]
        Nullable{T}()
    else
        Nullable{T}(X.values[I...])
    end
end

function Base.setindex!(X::NullableArray, v::Nullable, I::Int...)
    if isnull(v)
        X.isnull[I...] = true
    else
        X.isnull[I...] = false
        X.values[I...] = v
    end
    v
end

function Base.setindex!(X::NullableArray, v::Any, I::Int...)
    X.isnull[I...] = false
    X.values[I...] = v
    v
end

Base.similar(X::NullableArray, T, dims::Dims) = NullableArray(T, dims)

import Base: LinearFast

Base.linearindexing{T <: NullableArray}(::Type{T}) = LinearFast()

X = NullableArray(
    [1, 2, 3],
    [false, false, true]
)

X[1]

X[1:3]
```

The problem is that last line: as I understand it, you invoke `similar` and end up allocating a `NullableArray{Nullable{Int64},1}` that gets populated in terms of calls to the scalar `getindex` method, but I want a `NullableArray{Int64,1}` to be allocated instead.

I'm sure I've violated some assumptions you're making about how to determine the type `T` that gets passed to `similar`. What's the best way to resolve this?

mbauman commented:
`similar` is quite a beast.  It could use it's own section in an Interfaces manual page.  I didn't really understand it until I undertook this project, myself.

* `similar(A::AbstractArray)` falls back to `similar(A, size(A)`
* `similar(A::AbstractArray, dims::Dims)` falls back to `similar(A, eltype(A), dims)`
* `similar(A::AbstractArray, T, dims::Dims)` calls `Array(T, dims)` (after applying this PR)

You can intercept it at any point along the chain.  In this case, I'd *only* define:

```jl
Base.similar{T}(X::NullableArray{T}, dims::Dims) = NullableArray(eltype(T), dims)
```

If the user asks for a specific type in their call to `similar`, then I'd just use the Base fallback to `Array` since you can't do the Array-of-Structs to Struct-of-Arrays for an arbitrary eltype (it may not even be a `Nullable`).

After a minor fix in `setindex` (you need to unpack the Nullable with `get`: `X.values[I...] = get(v)`), all works as you'd hope:

```jl
julia> X = NullableArray(
           [1, 2, 3],
           [false, false, true]
       )
3-element NullableArray{Int64,1}:
 Nullable(1)
 Nullable(2)
 Nullable{Int64}()

julia> X[1]
Nullable(1)

julia> X[1:3]
3-element NullableArray{Int64,1}:
 Nullable(1)
 Nullable(2)
 Nullable{Int64}()
```

johnmyleswhite commented:
That's awesome. Thanks.

I'll think through exactly which `similar` methods we should expose. It's not clear to me why `similar` needs to care about `Nullable` or `eltype` when the user specifies a type `T` manually. I would think `similar(NullableArray, T, dims)` should produce a `NullableArray{T}`, no matter what `T` is.

I'm happy to write up an example like this one as documentation for this new feature.

StefanKarpinski commented:
@mbauman, if this work has lead you to feel that the `similar` interface could be improved in any way, that would be great. It's seems to work quite well, but it also evolved kind of organically over time. I love this whole change, btw.

mbauman commented:
I think part of why this is so tricky is because we want this struct of arrays to look and behave like an array of structs.  We want `NullableArray` to act *precisely* like an `Array{Nullable}`.

I suppose making the fallback to `Array` slightly changes the meaning of `similar`.  The current help states that it returns the same type except for "special objects" like Ranges.  I think that part of the doc update should include a rewrite of that.  Here's a draft:

> Base.similar(array, element_type, dims)

> Create an uninitialized and modifiable array with the specified element type and dimensions. The second and third arguments are both optional, defaulting to the properties of the passed array. The "dims" argument may be a tuple or a series of integer arguments. Custom AbstractArray types may choose to override the default behavior of returning an `Array` and instead return an uninitialized array of the same type as the passed array.

mbauman commented:
Thanks @StefanKarpinski. 

I think the similar interface actually works very well for this change.  The key part is that I think we should document that it's simply a way to ask for a chunk of array storage in a way that the AbstractArray type can say something about what kind of array it returns.

johnmyleswhite commented:
I think the tension is in this sentence:

    Create an uninitialized and modifiable array with the specified element type and dimensions.

In the case of `NullableArray`, I think you want the `T` in `similar` to be the first type parameter for `NullableArray{T, N}`, rather than the `eltype` of `NullableArray{T, N}`. For most `AbstractArray` types, there's a one-to-one mapping between first type parameter and eltype, but that doesn't apply to `NullableArray`.

mbauman commented:
I think I'd call that an implementation detail.  Generic code doesn't care what the first parameter of NullableArray is, but it definitely does care about the first parameter of AbstractArray.  If you are reporting to `AbstractArray{T}` that `T<:Nullable{S}` then `T` is your element type for all functions in the AbstractArray interface.

I think that if you want to work with an `S` then you *already know* that you need a NullableArray… so instead of `similar` you should be using the `NullableArray(T, dims)` constructor.

timholy commented:
Not to derail the important `similar` discussion, but can this be merged now? The Travis failures, are, I presume, the recent EOFError problem and not related to this PR?

johnmyleswhite commented:
Yeah, I think this PR is clearly ready to merge in my mind. My concerns are primarily relevant to documenting the assumptions this code makes so that it's easy for an end-user (like me) to figure out exactly which methods have to be implemented to get the benefits of this change.

StefanKarpinski commented:
The fact that it's even *possible* to enumerate those assumptions is a huge step forward as compared to the old (current) state of indexing, which was (is) a total morass.

StefanKarpinski commented:
@mbauman, whenever you're ready, please do the honors of merging this.

mbauman commented:
![SO FREAKING EXCITING](https://38.media.tumblr.com/3c63f816f976ea97fb9228312ea10433/tumblr_inline_np76uaiwVW1su4kc2_500.gif)

Will rebase and merge.

timholy commented:
:fireworks: 0.4 gets more awesome all the time.

Thanks a ton for this, @mbauman. Also to the unknown hero who fixed the codegen problems.

IainNZ commented:
:+1: lovely

kshyatt commented:
Julia super-:star2:s strike again!

JeffBezanson commented:
Very nice. :100:

It will be a fun challenge to try to cut down on staged functions, but that's a long term goal.

tlycken commented:
This might just be my favorite change in all of 0.4. I'm soo happy right now! 

@jeffbezanson The number of staged functions in packages such as Interpolations.jl will decrease by a factor of a handful, so I think that your long-term goal is already being worked on by this PR, only not in base :)

StefanKarpinski commented:
![slow clap orson welles](https://cloud.githubusercontent.com/assets/153596/7984791/8f7bf2d6-0a99-11e5-959c-bbaccf7561b4.gif)

Bravissimo!

ScottPJones commented:
:clap: :100:

ViralBShah commented:
:cake: :+1:

ViralBShah commented:
This may already be there (although not obvious with a quick scan). Do we have tests that will fail if we have a PR that breaks this amazing piece of work?

mbauman commented:
Somewhat. All base array types also fall back to this machinery, so I had been relying on their tests. But they do (all?) have at least some first-class methods. I'll add some simple tests for a barebones type.

mbauman commented:
Man, I'm worried that the Win32 codegen bug is currently hiding due to the unintentional inlining changes that caused a performance regression (which I really should have tested one last time after rebasing this).  I have a hunch that fixing #11595 will cause the codegen bug to reappear. :-\

yuyichao commented:
@mbauman  What is the win32 codegen bug? (issue number?)

tkelman commented:
No issue number for it. https://github.com/JuliaLang/julia/pull/10525#issuecomment-100420099

tkelman commented:
I can apply the patch from https://github.com/JuliaLang/julia/issues/11595#issuecomment-109515378 in a local win32 build and it doesn't cause the test failure to come back. We'll have to test with the other issue that's in #11595 too.

timholy commented:
Awesome. If you're willing, please test what happens when you revert ebfebfdc48919b767124a62958b9b6455552f3c9. That way we'll have a pretty good idea whether @mbauman's concern is worth worrying about even before the fix arrives.

tkelman commented:
Had the same thought. Looks fine locally with a revert.

timholy commented:
Whew.

tkelman commented:
If we really want to determine exactly what may have fixed the codegen problem, @mbauman did you have a branch where you rearranged things so it looks as if this PR was merged a month ago, then master replayed on top of that?

mbauman commented:
Yes, I do.  I just pushed it to my fork: https://github.com/JuliaLang/julia/compare/master...mbauman:mb/reorderedbisect

The commit SHAs have all changed, of course, so here's the bisect start/stop endpoints:

`git bisect start a8c1cd2cf68aa7dc7dc94bbb8556f773d45cc3a5 ccefc34abb883ce72ac687a751abc54679e68c5a`

mbauman commented:
Bisect is crazy powerful… just used it for the first time in hunting this stuff down.  Although its performance doesn't quite match [`git bisect run @timholy`](https://github.com/JuliaLang/julia/issues/11595#issuecomment-109607690).

tkelman commented:
Uh-oh, now I'm confused/worried. According to bisect (where good = reproduces codegen bug, bad = passes numbers test) on your branch, the first bad commit is indeed fdc0151a51e833c82c68cb4bb3a0de1a8d5c96a8. Now what is different between that and manually applying the reverse diff of ebfebfdc48919b767124a62958b9b6455552f3c9 on master?

mbauman commented:
Could it be an interaction between the inlining factor-of-1000 patch and this?  That is, when you reverted ebfebfd on master, did you *also* have the factor-of-1000 fix in?

tkelman commented:
Yes I did. But running 77226d2adeb4087a9471a67902d3c7512081a786 (which is the parent commit of fdc0151a51e833c82c68cb4bb3a0de1a8d5c96a8 on your alternate-history branch) with the factor-of-1000 patch applied still reproduces the codegen issue, so I think it's something else?

mbauman commented:
Very strange.  Thanks so much for digging into this.  Let's wait to see what happens once Jeff has time to address the `meta` issue.

tkelman commented:
It looks like there's a non-trivial difference between where this was merged and the recent end of that modified branch, checking `git diff 585fef06033d05728411ffbe2a39adaa37cb2a51 a8c1cd2cf68aa7dc7dc94bbb8556f773d45cc3a5`. Something missing or different from the rebase?

mbauman commented:
Very possible. I don't entirely know what the best way is to do this. I simply did an interactive rebase, adding the commits from mb/abstractsmarts10 at the right place by hand. There were a few "empty commits," which I didn't understand. It's possible I could try again and end up with a cleaner result.

tkelman commented:
I made a modified branch at https://github.com/tkelman/julia/tree/mb/bisect-10525-codegen where I inserted https://github.com/tkelman/julia/commit/72601525b88211b3c5373f2be4a131f0ebe0621c which covered the differences. I'm bisecting on that now and will see if it shows anything different.

tkelman commented:
Sorry, I really must've done something wrong the first time around. Reverting https://github.com/JuliaLang/julia/commit/ebfebfdc48919b767124a62958b9b6455552f3c9, with or without the change to the inlining factor, does bring the codegen problem back.

timholy commented:
Well, perhaps now it will get more attention. It's not the only 32bit problem we have, see #11026.

tkelman commented:
I would hope so, but so far my requests for ideas of other things to try in this function https://github.com/JuliaLang/julia/blob/00e9babb726c64507689843cc9454e19182f2d1e/base/intfuncs.jl#L285 have gone unanswered. Help us @JeffBezanson, you're our only hope.

Reminder to myself and others that next time a bug mysteriously goes away and we don't know why, maybe spend a bit more time looking into it.

We should probably open a separate issue for debugging this though, since this one's closed and getting overly long. Master doesn't exhibit the bug right now, but fixing #11595 would likely cause it to come back.

vtjnash commented:
@tkelman I posted some more info to your gist (I forgot GitHub doesn't ping automatically). 

@keno since you mentioned this may be a bug in stackrestore, I've started work on a branch to remove it

tkelman commented:
Oh, thanks, replied there - https://gist.github.com/anonymous/17f8c8b2a3212942d43d#comment-1468609

gists are poorly suited for discussion though due to the lack of notifications.
